client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -0,0 +1,262 @@
+        }
+    }
+
+    @Override
+    public List<Cookie> getAll() {
+        try {
+            lock.lock();
+            final boolean[] removeExpired = {false};
+            List<Cookie> result = cookieJar
+                    .entrySet()
+                    .stream()
+                    .filter(pair -> {
+                        boolean hasCookieExpired = hasCookieExpired(pair.getValue().cookie, pair.getValue().createdAt);
+                        if (hasCookieExpired && !removeExpired[0])
+                            removeExpired[0] = true;
+                        return !hasCookieExpired;
+                    })
+                    .map(pair -> pair.getValue().cookie)
+                    .collect(Collectors.toList());
+
+            if (removeExpired[0])
+                removeExpired();
+
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean remove(Cookie cookie) {
+        try {
+            lock.lock();
+            return cookieJar.entrySet().removeIf(v -> v.getValue().cookie == cookie);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean clear() {
+        try {
+            lock.lock();
+            boolean result = !cookieJar.isEmpty();
+            cookieJar.clear();
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    private String requestDomain(Uri requestUri) {
+        return requestUri.getHost().toLowerCase();
+    }
+
+    private String requestPath(Uri requestUri) {
+        return requestUri.getPath().isEmpty() ? "/" : requestUri.getPath();
+    }
+
+    // rfc6265#section-5.2.3
+    // Let cookie-domain be the attribute-value without the leading %x2E (".") character.
+    private AbstractMap.SimpleEntry<String, Boolean> cookieDomain(String cookieDomain, String requestDomain) {
+        if (cookieDomain != null) {
+            String normalizedCookieDomain = cookieDomain.toLowerCase();
+            return new AbstractMap.SimpleEntry<>(
+                    (!cookieDomain.isEmpty() && cookieDomain.charAt(0) == '.') ?
+                            normalizedCookieDomain.substring(1) :
+                            normalizedCookieDomain, false);
+        } else
+            return new AbstractMap.SimpleEntry<>(requestDomain, true);
+    }
+
+    // rfc6265#section-5.2.4
+    private String cookiePath(String rawCookiePath, String requestPath) {
+        if (MiscUtils.isNonEmpty(rawCookiePath) && rawCookiePath.charAt(0) == '/') {
+            return rawCookiePath;
+        } else {
+            // rfc6265#section-5.1.4
+            if (!requestPath.isEmpty() && requestPath.charAt(0) == '/' && requestPath.lastIndexOf('/') > 0)
+                return requestPath.substring(0, requestPath.lastIndexOf('/'));
```

Comment:
Not exactly my commit: you're computing lastIndexOf('/') twice :)

Reply:
Crap, my bad. I pushed the fix.